### PR TITLE
*Syntactically* permit visibilities on trait items & enum variants

### LINKED ISSUE #65041: an empty `:vis` doesn't work in trait declaration
I want to create a macro that generate `fn`s that use in both `impl Item...` and `trait TraitName`. In `trait`, `pub` must be omitted, but in `impl SomeStruct` I need `pub`, so the macro need a `$vis:vis` like this:

```rust
macro_rules! create_method {
    ($vis:vis $name:ident) => {
        $vis fn $name(&self) {
            println!("{}", stringify!($name));
        }
    }
}

trait T1 {
    // This not works
    create_method!(method_of_t1);
}

trait T2 {
    fn method_of_t2(&self);
}

struct Struct;

impl T2 for Struct {
    // This works
    create_method!(method_of_t2);
}

impl Struct {
    // I want this method to be `pub`, so the macro need a `:vis`
    // for receiving the visibility keyword like this:
    create_method!(pub another_method);
}
```
Expected result: `create_method!(method_of_t1);` in `trait T1` should works.

(**Edit**: a better comment about the `:vis` in `impl Struct`)

### PR DESCRIPTION
**Pull request by Centril**
Fixes #65041

Suppose we have `$vis trait_item` or `$vis enum_variant` and `$vis` is a `:vis` macro fragment. Before this PR, this would fail to parse. This is now instead allowed as per language team consensus in https://github.com/rust-lang/rust/issues/65041#issuecomment-538105286. (See added tests for elaboration.)

Moreover, we now also permit visibility modifiers on trait items & enum variants *syntactically* but reject them with semantic checks (in `ast_validation`):

```rust
#[cfg(FALSE)]
trait Foo { pub fn bar(); } // OK

#[cfg(FALSE)]
enum E { pub U } // OK
```

### CODE REVIEW

File: src/libsyntax/parse/parser.rs [hunk 1]
```diff
@@ -1205,6 +1205,23 @@ impl<'a> Parser<'a> {
         Ok(())
     }
 
+    /// Parses an inherited visibility (`VisibilityKind::Inherited`)
+    /// or errors with parser recovery. See #65041 for more context.
+    fn parse_visibility_inherited(&mut self) -> PResult<'a, ()> {
+        let snapshot_expected = self.expected_tokens.clone();
```
Comment (Centril) [inline src/libsyntax/parse/parser.rs]: Maybe we should instead do:
```suggestion
        let snapshot_expected = mem::take(&mut self.expected_tokens);
```
which will likely allocate again, but probably less work than `.clone()` overall.

File: src/test/ui/parser/mismatched-braces/missing-close-brace-in-trait.stderr [hunk 2]
```diff
@@ -1,25 +1,28 @@
...
-   |                   |
-   |                   expected one of 7 possible tokens here
-   |                   help: `}` may belong here
-LL | 
+...
 LL | pub(crate) struct Bar<T>();
-   | ^^^ unexpected token
+   |           -^^^^^^ expected one of 7 possible tokens here
+   |           |
+   |           help: `}` may belong here
```
Comment (estebank) [inline src/test/ui/parser/mismatched-braces/missing-close-brace-in-trait.stderr]: This is... unfortunate.

**Comment by estebank** [comment]
LGTM, but saw that @petrochenkov self-assigned, letting him sign off on it.

**Comment by petrochenkov** [comment]
So, I agree with the (direction of) the change for trait items, because that's a part of the syntactic unification of free/trait/impl/foreign items.
The parser should accept any visibilities on them (inherited or not), and then AST validation will produce semantic errors if necessary. (If that requires AST changes that you don't want to do in this PR, then it can be reported in the parser for now.)

I don't agree with accepting empty `vis` on variants, or at least don't feel confident about it.
Even in the token-based model `vis` is rather a delimited group with an empty token stream inside it rather than an empty token stream literally and I'm not sure we should accept it unless we accept visibilities on variants in general (syntactically).
We can certainly parse visibilities on variants for recovery though (while always producing an error).

File: src/libsyntax/parse/parser.rs (hunk 1 above)
Comment (Centril) [reply src/libsyntax/parse/parser.rs]: This is moot now.

**Comment by Centril** [comment]
@petrochenkov Adjusted the PR to *syntactically* accept any visibility on enum variants & trait items now (discussion in https://rust-lang.zulipchat.com/#narrow/stream/213817-t-lang/topic/.2366183 -- cc @rust-lang/lang again in case anyone missed it)

**Comment by petrochenkov** [comment]
LGTM, but needs commit squashing and a pprust test update.

**Comment by petrochenkov** [comment]
Let's pass this through an FCP, I'm not sure the enum variant change has enough motivation and thus don't want to take responsibility for it.

**Comment by Centril** [comment]
Alright; let's do that.

Dear language team & community, I would like to propose that we *syntactically* (but not semantically) permit visibility modifiers on items in `trait Foo { ... }` and on `enum Foo { pub Variant }`. In the case of trait items, this will facilitate the merging of the various item forms syntactically and thereby simplify the grammar and the compiler. As for enum variants, I believe the change improves diagnostics recovery, could facilitate macro based DSLs, and eventually could lead to allowing visibility modifiers semantically as well.

@rfcbot merge

File: src/test/ui/parser/mismatched-braces/missing-close-brace-in-trait.stderr [hunk 3]
```diff
@@ -1,25 +1,30 @@
...
+  --> $DIR/missing-close-brace-in-trait.rs:1:1
+   |
+LL | / trait T {
+LL | |
+LL | |     fn foo(&self);
+LL | |
+...  |
+LL | |
+LL | | fn main() {}
+   | |_________________________________________________________________^ consider adding a `main` function to `$DIR/missing-close-brace-in-trait.rs`
```
Comment (estebank) [inline src/test/ui/parser/mismatched-braces/missing-close-brace-in-trait.stderr]: I'm surprised this error is emitted, as when there are unclosed delimiters it shouldn't :-/

**Comment by Centril** [comment]
@rfcbot reviewed

(please wake up... ^^)

**Comment by petrochenkov** [comment]
@bors r+
(People have ticked their boxes, and this is a tiny feature, not much reason to wait for an extra week.)

**Comment by XAMPPRocky** [comment]
@Centril What specifically about this should be mentioned in the release notes? For practical purposes nothing has really changed on the user facing side as far as I can tell, since they will get an error no matter what.

**Comment by Centril** [comment]
@XAMPPRocky Most of the time nothing has changed for users. However, in niche circumstances, e.g. with a procedural macro, you can do e.g.:
```rust
#[my_funky_attr]
enum Foo { pub Bar }

// ...expands to the following because `my_funky_attr` is funky:

struct Bar;
enum Foo { Bar(Bar) }
```

The same logic applies to trait items.

**Comment by dtolnay** [comment]
Concrete use case: DSL for a Java-style enum (https://docs.oracle.com/javase/tutorial/java/javaOO/enum.html).

Java:

```java
public enum Planet {
    MERCURY (3.301e+23, 2.440e+6),
    VENUS   (4.868e+24, 6.052e+6),
    EARTH   (5.972e+24, 6.371e+6),
    MARS    (6.417e+23, 3.390e+6),
    JUPITER (1.898e+27, 6.991e+7),
    SATURN  (5.683e+26, 5.823e+7),
    URANUS  (8.681e+25, 2.536e+7),
    NEPTUNE (1.024e+26, 2.462e+7);

    private final double mass;
    private final double radius;

    Planet(double mass, double radius) {
        this.mass = mass;
        this.radius = radius;
    }

    // additional methods
}
```

Rust:

```rust
#[aux_enum_data]
pub enum Planet {
    Mercury = (3.301e+23, 2.440e+6),
    Venus   = (4.868e+24, 6.052e+6),
    Earth   = (5.972e+24, 6.371e+6),
    Mars    = (6.417e+23, 3.390e+6),
    Jupiter = (1.898e+27, 6.991e+7),
    Saturn  = (5.683e+26, 5.823e+7),
    Uranus  = (8.681e+25, 2.536e+7),
    Neptune = (1.024e+26, 2.462e+7),

    pub PlanetFacts {
        pub mass: f64,
        pub radius: f64,
    }
}
```

Expands to:

```rust
#[derive(Copy, Clone, PartialEq, Eq, Hash)]
pub enum Planet {
    Mercury,
    Venus,
    Earth,
    Mars,
    Jupiter,
    Saturn,
    Uranus,
    Neptune,
}

pub struct PlanetFacts {
    pub mass: f64,
    pub radius: f64,
}

impl std::ops::Deref for Planet {
    type Target = PlanetFacts;

    fn deref(&self) -> &Self::Target {
        match self {
            Planet::Mercury => &PlanetFacts { mass: 3.301e+23, radius: 2.440e+6 },
            Planet::Venus   => &PlanetFacts { mass: 4.868e+24, radius: 6.052e+6 },
            Planet::Earth   => &PlanetFacts { mass: 5.972e+24, radius: 6.371e+6 },
            Planet::Mars    => &PlanetFacts { mass: 6.417e+23, radius: 3.390e+6 },
            Planet::Jupiter => &PlanetFacts { mass: 1.898e+27, radius: 6.991e+7 },
            Planet::Saturn  => &PlanetFacts { mass: 5.683e+26, radius: 5.823e+7 },
            Planet::Uranus  => &PlanetFacts { mass: 8.681e+25, radius: 2.536e+7 },
            Planet::Neptune => &PlanetFacts { mass: 1.024e+26, radius: 2.462e+7 },
        }
    }
}
```

```rust
println!("1 earth mass = {} kg", Earth.mass);
```

**Pull request merged**